### PR TITLE
Add ProfilerHook for capturing CPU/GPU profiling information with MonitoredSession

### PR DESCRIPTION
**Pull request by darrengarvey**
It would be good to see if this is the preferred way to get profiling when using tf.contrib.training. If so, I can add the missing tests and doc updates.

### CODE REVIEW

**Comment by drpngx** [comment]
@ispirmustafa could you take a look?

**Comment by yaroslavvb** [comment]
This is pretty useful, I was wondering how to do the same thing myself

File: tensorflow/python/training/basic_session_run_hooks.py
Comment (ispirmustafa) [inline tensorflow/python/training/basic_session_run_hooks.py]: This seems to be a useful hook. Could you please add tests?

**Comment by darrengarvey** [comment]
I've added tests and updated the docs.

One test does have a sleep in it which isn't great, but I don't see an existing utility (eg. faking time) to get around that. The test does at least try to follow the existing conventions in that file.

**Comment by darrengarvey** [comment]
Excellent. I spotted @gunan just put in the use of `@test.mock.patch`, which I gladly used. The ProfilerHook test now takes < 1s rather than nearly 7s, which is a bit more like it. Thanks.

I also fixed up the merge conflict.

**Comment by caisq** [comment]
@tensorflow-jenkins test this please
@ispirmustafa PTAAL?

File: tensorflow/python/training/basic_session_run_hooks.py
Comment (gunan) [inline tensorflow/python/training/basic_session_run_hooks.py]: Lint error here, please indent by 2 after with.

File: tensorflow/python/training/basic_session_run_hooks_test.py
Comment (gunan) [inline tensorflow/python/training/basic_session_run_hooks_test.py]: Lint error here, please indent by 2 spaces.

**Review by gunan — CHANGES_REQUESTED** [review]

**Comment by darrengarvey** [comment]
Thanks for the review @gunan. I've fixed the lines you spotted, but haven't figured out yet how to lint the TF code so hopefully it'll pass now:

I tried:

    bazel test -c opt --copt="-march=native" --config=cuda //tensorflow/python:basic_session_run_hooks_test

and:

    PYLINTRC=./tensorflow/tools/ci_build/pylintrc pylint tensorflow/python/training/basic_session_run_hooks_test.py

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by yaroslavvb** [comment]
btw, to lint things:

pip install pylint
wget -O /tmp/pylintrc
https://raw.githubusercontent.com/tensorflow/tensorflow/master/tensorflow/tools/ci_build/pylintrc
pylint tensorflow/get_string_ops.py --rcfile=/tmp/pylintrc

File: tensorflow/contrib/learn/python/learn/basic_session_run_hooks.py
Comment (ispirmustafa) [inline tensorflow/contrib/learn/python/learn/basic_session_run_hooks.py]: you don't need to put this line here. this file is obsolete.

File: tensorflow/python/training/basic_session_run_hooks_test.py
Comment (ispirmustafa) [inline tensorflow/python/training/basic_session_run_hooks_test.py]: Can you use SingularMonitoredSession instead of test_session? which will call begin and before/after run automatically.

Comment (darrengarvey) [inline tensorflow/python/training/basic_session_run_hooks_test.py]: Thanks for the tip. That's cleaner now.

File: tensorflow/contrib/learn/python/learn/basic_session_run_hooks.py
Comment (darrengarvey) [inline tensorflow/contrib/learn/python/learn/basic_session_run_hooks.py]: Good to know. Fixed.

**Comment by drpngx** [comment]
@ispirmustafa @gunan are we good with the amended commit?

**Comment by gunan** [comment]
I am ok, but Id like @ispirmustafa to take one final look at the PR.
Jenkins, test this please.

**Comment by darrengarvey** [comment]
Thanks for the reviews.

The Windows CMake test [failed](https://ci.tensorflow.org/job/tensorflow-pr-win-cmake-py/896/consoleFull), but that looks like an unrelated race condition in the [input_test.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/input_test.py#L987). A vague look seems like ``all_a`` and ``seen_b`` should be getting checked rather than ``which_a`` and ``which_b``, [possibly](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/input_test.py#L979).

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by ispirmustafa** [comment]
I'm good with the PR.
Thanks @darrengarvey!

**Review by gunan — APPROVED** [review]

**Comment by gunan** [comment]
I think this involves an API change. Could you comment about the change @martinwicke

**Review by drpngx — CHANGES_REQUESTED** [review]
As per @gunan, this introduces an API change, so let's not merge this just now.

**Comment by martinwicke** [comment]
​Yes. We will get a decision next Tuesday. Should be fine.

**Comment by darrengarvey** [comment]
Thanks for the reviews all.

I'll leave it up to you guys. Let me know if I should update the PR to not expose it to the public API and call the class `_ProfilerHook` like I've noticed other hooks - like `FeedFnHook` - did while they "bed in".

**Comment by martinwicke** [comment]
Yes, can you move this to contrib for now? You can make a new `contrib/hooks` of if you have another good spot it fits into, that would be fine too.

**Comment by rmlarsen** [comment]
@darrengarvey @martinwicke it looks like we are waiting for this to be moved.

**Comment by darrengarvey** [comment]
I've had a go at moving it to contrib.hooks. I couldn't find an easier place to put it.

**Comment by yaroslavvb** [comment]
BTW, locations in contrib mirror official locations for easy of moving between the two, so

`<gitroot>/tensorflow/<somepath>` is translated to `<gitroot>/tensorflow/projectname/<somepath>`

So in your case "hooks" is project name, but `<somepath>` part remains the same. https://github.com/tensorflow/tensorflow/commit/37fbebdd is a recent example of moving something to contrib

**Comment by darrengarvey** [comment]
Thanks for the pointers @yaroslavvb, but I'm not sure I entirely follow. :) Do you mean:

`<gitroot>/tensorflow/<somepath>` is translated to `<gitroot>/tensorflow/contrib/projectname/<somepath>`

(note the contrib)?

Did you notice I had pushed my attempt to the same branch? I have only moved `ProfilerHook` which was previously alongside the other hooks in `basic_session_run_hooks.py` so I had to give it a new file. I don't mind moving the whole of `basic_session_run_hooks.py` to `contrib/hooks`, but since the other hooks are already available I didn't take @martinwicke's comment to mean that.

I don't mind getting this in the right place; I'm getting quite a lot back from TF so I've got a fair amount of patience. :)

**Comment by yaroslavvb** [comment]
Yes, that's the path I meant, updated. Correct, I don't think Martin meant for you to move existing functionality to `contrib`, just the new addition.

For example of existing placement consider locations of existing hooks:

Official "basic_session_run_hooks"
`tensorflow/python/training/basic_session_run_hooks.py`

Unofficial hooks that should at some point be merged into file above
`tensorflow/contrib/learn/python/learn/basic_session_run_hooks.py `

Currently your path is `tensorflow/contrib/hooks/python/hooks/profiler_hook.py` which means that for integration it would be moved to `tensorflow/python/hooks/profiler_hook.py`, which would create a new package `python/hooks` in the TensorFlow repo.

To keep hooks under `python/learn`, you could use following locations

1. Add to existing `learn` project, modify `basic_session_run_hooks.py`
`tensorflow/contrib/learn/python/learn/basic_session_run_hooks.py `

2. Move to `hooks` project, reuse `basic_session_run_hooks.py` file
`tensorflow/contrib/hooks/python/learn/basic_session_run_hooks.py `

3. Move to `hooks` project, use new file
`tensorflow/contrib/hooks/python/learn/profiler_hook.py `

I think your current solution is closest to 3.

**Comment by martinwicke** [comment]
I have some preference for making this a `contrib/hooks` project. I think the hooks should probably live in the directory `python/training` once it's in core, next to `MonitoredSession`.

So, the path to use would be: `contrib/hooks/python/training/profiler_hook.py`. I think we want to have one file per hook in the end, so no need to use `basic_run_hook.py`.

**Comment by darrengarvey** [comment]
Ah I hadn't picked up on that mirroring magic. Good to know.

I'm away for the weekend but I'll do this tweak to put it under `contrib/hooks/python/training/` probably on Monday.

**Comment by darrengarvey** [comment]
Right, pushed as per suggestions, plus a couple of fixes. The hooks aren't specifically for ``contrib.learn``, so putting them under ``contrib/hooks/python/training/`` made sense.

File: tensorflow/python/training/basic_session_run_hooks.py
Comment (gunan) [inline tensorflow/python/training/basic_session_run_hooks.py]: Since all the other changes are reverted in this file, do we need these modifications?

File: tensorflow/python/training/training.py
Comment (gunan) [inline tensorflow/python/training/training.py]: leftover empty line?

File: tensorflow/contrib/hooks/BUILD
Comment (gunan) [inline tensorflow/contrib/hooks/BUILD]: This rule also needs to be referenced from `all_opensource_files` rule in `tensorflow/BUILD` file.
Otherwise it will be excluded from our pip packages.

File: tensorflow/python/training/basic_session_run_hooks.py
Comment (darrengarvey) [inline tensorflow/python/training/basic_session_run_hooks.py]: No. Also removed other lines I added previously.

File: tensorflow/python/training/training.py
Comment (darrengarvey) [inline tensorflow/python/training/training.py]: Yep, removed.

File: tensorflow/contrib/hooks/BUILD
Comment (darrengarvey) [inline tensorflow/contrib/hooks/BUILD]: Ah, I'd checked it worked from ``python setup.py develop``. I'll try the full pip install next time. Thanks.

**Comment by darrengarvey** [comment]
I'm not sure if it's OK for me to ask Jenkins to test this, to check I didn't miss anything?

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Could you check the test? It might be related.

```
15:44:05 151/201 Test #168: C:/tf_jenkins/home/workspace/tensorflow-pr-win-cmake-py/tensorflow/python/training/coordinator_test.py .........................................***Failed   15.53 sec
15:44:05 ....Exception in thread Thread-15:
15:44:05 Traceback (most recent call last):
15:44:05   File "C:\Program Files\Anaconda3\lib\threading.py", line 914, in _bootstrap_inner
15:44:05     self.run()
15:44:05   File "C:\Program Files\Anaconda3\lib\threading.py", line 862, in run
15:44:05     self._target(*self._args, **self._kwargs)
15:44:05   File "C:/tf_jenkins/home/workspace/tensorflow-pr-win-cmake-py/tensorflow/python/training/coordinator_test.py", line 39, in RaiseInN
15:44:05     raise ex
15:44:05 ValueError: Clean stop
15:44:05 
15:44:05 .Exception in thread Thread-16:
15:44:05 Traceback (most recent call last):
15:44:05   File "C:\Program Files\Anaconda3\lib\threading.py", line 914, in _bootstrap_inner
15:44:05     self.run()
15:44:05   File "C:\Program Files\Anaconda3\lib\threading.py", line 862, in run
15:44:05     self._target(*self._args, **self._kwargs)
15:44:05   File "C:/tf_jenkins/home/workspace/tensorflow-pr-win-cmake-py/tensorflow/python/training/coordinator_test.py", line 39, in RaiseInN
15:44:05     raise ex
15:44:05 tensorflow.python.framework.errors_impl.OutOfRangeError: First
15:44:05 
15:44:05 .F............
15:44:05 ======================================================================
15:44:05 FAIL: testJoinRaiseReportExcInfo (__main__.CoordinatorTest)
15:44:05 ----------------------------------------------------------------------
15:44:05 RuntimeError: Too late
15:44:05 
15:44:05 During handling of the above exception, another exception occurred:
15:44:05 
15:44:05 Traceback (most recent call last):
15:44:05   File "C:/tf_jenkins/home/workspace/tensorflow-pr-win-cmake-py/tensorflow/python/training/coordinator_test.py", line 162, in testJoinRaiseReportExcInfo
15:44:05     coord.join(threads)
15:44:05 AssertionError: "First" does not match "Too late"
15:44:05 
15:44:05 ----------------------------------------------------------------------
15:44:05 Ran 19 tests in 6.751s
15:44:05 
15:44:05 FAILED (failures=1)
```

**Comment by darrengarvey** [comment]
Thanks @drpngx. ``coordinator_test.py`` failed, which looks unrelated. It's after midnight here so I'll try have a look tomorrow.

**Comment by darrengarvey** [comment]
The failure was certainly unrelated.

https://github.com/tensorflow/tensorflow/blob/58201a058853de647b37ddb0ccf63d89b2357f03/tensorflow/python/training/coordinator_test.py#L152

That test looks like it'll fail under heavy load - ie. when starting a thread #1 takes more than 40ms more than starting thread #2. I can't replicate the failure here under heavy load but I might try it again with some more exotic cgroup settings if I manage to get time, but that's not likely this week. The test just below looks equally susceptible to thread starting times.

The same time mocking trick we did here should work for this test, but that's for another PR AFAICT.

**Comment by gunan** [comment]
Just for the peace of mind
Jenkins, test this please.

**Comment by darrengarvey** [comment]
Hmm, a link-time failure on Linux cmake builds this time. Gadzooks... I'll have a better look in the morning, but for reference:

https://ci.tensorflow.org/job/tensorflow-pull-requests-cpu-cmake/1610/console

``
[98%] Linking CXX executable compare_graphs
CMakeFiles/tf_core_kernels.dir/workspace/tensorflow/core/kernels/hexagon/graph_transfer_utils.cc.o: In function `tensorflow::GraphTransferUtils::BuildFusedGraphDef(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<tensorflow::GraphTransferer::InputNodeInfo, std::allocator<tensorflow::GraphTransferer::InputNodeInfo> > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, tensorflow::GraphDef const&, tensorflow::GraphTransferer*)':
graph_transfer_utils.cc:(.text+0x7c8): undefined reference to `tensorflow::Scope::NewRootScope()'
graph_transfer_utils.cc:(.text+0x88f): undefined reference to `tensorflow::Scope::WithOpName(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
graph_transfer_utils.cc:(.text+0x8d4): undefined reference to `tensorflow::Scope::GetUniqueNameForOp(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
graph_transfer_utils.cc:(.text+0x10ba): undefined reference to `tensorflow::Scope::UpdateBuilder(tensorflow::NodeBuilder*) const'
graph_transfer_utils.cc:(.text+0x10e0): undefined reference to `tensorflow::Scope::UpdateStatus(tensorflow::Status) const'
graph_transfer_utils.cc:(.text+0x112e): undefined reference to `tensorflow::Operation::Operation(tensorflow::Node*)'
graph_transfer_utils.cc:(.text+0x15b9): undefined reference to `tensorflow::Scope::WithOpName(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
graph_transfer_utils.cc:(.text+0x19b8): undefined reference to `tensorflow::ops::AsNodeOutList(tensorflow::Scope const&, tensorflow::InputList const&)'
graph_transfer_utils.cc:(.text+0x1aef): undefined reference to `tensorflow::Scope::GetUniqueNameForOp(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
graph_transfer_utils.cc:(.text+0x236a): undefined reference to `tensorflow::Scope::UpdateBuilder(tensorflow::NodeBuilder*) const'
graph_transfer_utils.cc:(.text+0x239b): undefined reference to `tensorflow::Scope::UpdateStatus(tensorflow::Status) const'
graph_transfer_utils.cc:(.text+0x23f8): undefined reference to `tensorflow::Scope::ToGraphDef(tensorflow::GraphDef*) const'
collect2: error: ld returned 1 exit status
make[2]: *** [compare_graphs] Error 1
CMakeFiles/compare_graphs.dir/build.make:1470: recipe for target 'compare_graphs' failed
CMakeFiles/Makefile2:3735: recipe for target 'CMakeFiles/compare_graphs.dir/all' failed
make[1]: *** [CMakeFiles/compare_graphs.dir/all] Error 2
``

**Comment by yaroslavvb** [comment]
I've been averaging 1-3 unrelated failing tests for my PR submissions. Back when I was on TensorFlow team I tried to make sure that all tests were 100% passing all the time but it was hard. You can make internal changes blocking on all tests passing only if tests are super-fast. But the tests are only super-fast if people take extra care to make their tests efficient/parallelized. So eventually end-to-end testing time gets slow and by the time failure is noticed, the person who broke them went to lunch/home/etc

**Comment by gunan** [comment]
Both failures are known issues.
We have fixes available for both of them, just waiting for merges.
https://ci.tensorflow.org/job/tensorflow-master-cpu-cmake/595/console

**Comment by yaroslavvb** [comment]
@gunan this is ready for merging, right?

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Just to confirm, this is an API change so I'd like @ispirmustafa to confirm it's OK.

**Comment by martinwicke** [comment]
This is fine to merge. The added API is in contrib, so no problem.

**Comment by drpngx** [comment]
Oh, right, good to go then.

**Review by drpngx — APPROVED** [review]

**Pull request merged**
